<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Agrupamiento de Datos

En este laboratorio usaremos el DataFrame creado en la lección anterior resultado de la unión de los conjuntos de datos compartidos contigo. 

**Instrucciones:**

1. Lee cuidadosamente el enunciado de cada ejercicio.

2. Implementa la solución en la celda de código proporcionada.

3. Devuelve los resultados siempre en formato de DataFrame

In [1]:
# Carga el dataset de trabajo
import pandas as pd
import numpy as np

df_Netflix = pd.read_csv ("datos/netflix_merged.csv")
df_Netflix.head()

Unnamed: 0            Title                  Genre           Premiere  \
0           0  Enter the Anime            Documentary     August 5, 2019   
1           1      Dark Forces               Thriller    August 21, 2020   
2           2          The App  Science fiction/Drama  December 26, 2019   
3           3   The Open House        Horror thriller   January 19, 2018   
4           4      Kaali Khuhi                Mystery   October 30, 2020   

   Runtime  IMDB Score          Language show_id   type            title  \
0       58         2.5  English/Japanese   s3624  Movie  Enter the Anime   
1       81         2.6           Spanish   s2105  Movie      Dark Forces   
2       79         2.6           Italian   s3095  Movie          The App   
3       94         3.2           English   s5074  Movie   The Open House   
4       90         3.4             Hindi   s1778  Movie      Kaali Khuhi   

                    director  \
0              Alex Burunova   
1          Bernardo Arellano   
2               Elisa Fuksas   
3  Matt Angel, Suzanne Coote   
4            Terrie Samundra   

                                                cast                country  \
0  Kozo Morishita, Yoko Takahashi, Shinji Aramaki...   United States, Japan   
1  Tenoch Huerta, Eréndira Ibarra, Dale Carley, A...                 Mexico   
2  Vincenzo Crea, Jessica Cressy, Greta Scarano, ...                  Italy   
3  Dylan Minnette, Piercey Dalton, Patricia Bethu...  Canada, United States   
4  Shabana Azmi, Leela Samson, Sanjeeda Sheikh, S...                  India   

          date_added  release_year rating duration  \
0     August 5, 2019          2019  TV-MA   59 min   
1    August 21, 2020          2020  TV-MA   81 min   
2  December 26, 2019          2019  TV-MA   79 min   
3   January 19, 2018          2018  TV-MA      NaN   
4   October 30, 2020          2020  TV-14      NaN   

                                           listed_in  \
0                Documentaries, International Movies   
1  Horror Movies, Independent Movies, Internation...   
2   Dramas, Independent Movies, International Movies   
3                           Horror Movies, Thrillers   
4                Horror Movies, International Movies   

                                         description  
0  What is anime? Through deep-dives with notable...  
1  In search of his sister, a renegade criminal s...  
2  While in Rome to shoot his first movie, actor ...  
3  Following a tragedy, a mother and her teen son...  
4  When a restless spirit curses a village that h...

#### Ejercicio 1: ¿Cuántos títulos hay por cada tipo de contenido ('Movie' o 'TV Show')?

In [13]:
df_ej = df_Netflix.groupby('type').size().reset_index(name='count')
df_ej


type  count
0  Movie    513

#### Ejercicio 2: ¿Cuál es el promedio de puntuación de las películas agrupadas por clasificación de audiencia ('rating')? Interpreta los resultados, ¿hay algo que te llame la atención? Explica por qué. 

In [10]:
condicion = df_Netflix['type'] == 'Movie'
df_calificacion = df_Netflix[condicion].groupby('rating')['IMDB Score']
df_promedio = df_calificacion.mean().reset_index()
df_promedio


rating  IMDB Score
0      G    6.500000
1     PG    6.800000
2  PG-13    6.386957
3      R    6.495745
4  TV-14    6.177660
5   TV-G    6.156250
6  TV-MA    6.097638
7  TV-PG    6.308772
8   TV-Y    6.780000
9  TV-Y7    6.240000

Observamos que al agrupar las películas por clasificación de audiencia y rating se muestra que las películas con rating PG tienen una media mayor que 6.8, mientras que las de rating TV-14 tienen el promedio más bajo con 6.18. Tal vez esto indique que las peliculas PG dirigidas a un público más amplio y familiar tiedan a recibir mejor calificaciónque las de rating más restrictivo

#### Ejercicio 3: ¿Cuántos títulos hay por cada país? Ordena los resultados de mayor a menor

In [25]:
df_por_pais=df_Netflix.groupby('country')
df_conteo_por_pais=df_por_pais.size().reset_index(name='count')
df_conteo_por_pais_ordenado = df_conteo_por_pais.sort_values(by='count', ascending=False)
df_conteo_por_pais_ordenado


country  count
67                           United States    274
22                                   India     35
59                          United Kingdom     13
30                                   Italy     13
50                                   Spain     12
..                                     ...    ...
73                  United States, Germany      1
75                    United States, Japan      1
78   United States, United Kingdom, Canada      1
79  United States, United Kingdom, Germany      1
80    United States, United Kingdom, India      1

[81 rows x 2 columns]

#### Ejercicio 4: ¿Cuál es la calificación promedio en IMDb de los títulos agrupados por tipo ('Movie' o 'TV Show')?

In [28]:
df_agrupado=df_Netflix.groupby('type')['IMDB Score']
df_promedio=df_agrupado.mean().reset_index()
df_promedio

type  IMDB Score
0  Movie    6.210916

#### Ejercicio 5: ¿Cuántos títulos se añadieron a Netflix cada año? Para resolver este ejercicio tendrás uqe convertir la columna `date_added` a formato fecha usando el método `pd.to_datetime`. [Aquí](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) tienes la documentación, leeala con detenimiento antes de usar el método. 

In [33]:
df_Netflix['date_added'] = pd.to_datetime(df_Netflix['date_added'])
df_Netflix['date_added']
# Vamos a extraer el año en que se añadieron los títulos a Netflix
df_Netflix['year_added'] = df_Netflix['date_added'].dt.year #encontrado en documentación de Pandas.series
df_por_anio=df_Netflix.groupby('year_added')
df_conteo_por_anio = df_por_anio.size().reset_index(name='count')
df_conteo_ordenado=df_conteo_por_anio.sort_values(by='year_added')
df_conteo_ordenado


year_added  count
0        2015     10
1        2016     28
2        2017     57
3        2018     83
4        2019    107
5        2020    164
6        2021     64

#### Ejercicio 6: ¿Cuál es el país con más títulos lanzados en el año 2020? Para resolver este ejercicio después del gropuby os puede resultar interesante el método [`idxmax`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html)

In [46]:
condicion = df_Netflix['release_year'] == 2020
df_lanzados_2020 = df_Netflix[condicion]
# Agrupamos por país y la condicion que hemos establecido  y hacemos count
df_conteo=df_lanzados_2020.groupby('country').size().reset_index(name='count')
pais_con_mas_titulos= df_conteo.loc[df_conteo['count'].idxmax()]
pais_con_mas_titulos





country    United States
count                 74
Name: 34, dtype: object

#### Ejercicio 7:¿Cuál es el promedio de puntuación en IMDb de los títulos agrupados por país?

In [49]:
df_media_por_pais = df_Netflix.groupby('country')['IMDB Score'].mean().reset_index()
df_media_por_pais

country  IMDB Score
0                                Argentina    6.420000
1                 Australia, United States    7.600000
2                                  Austria    6.850000
3                     Belgium, Netherlands    7.100000
4   Belgium, United Kingdom, United States    5.300000
..                                     ...         ...
76                   United States, Mexico    5.666667
77           United States, United Kingdom    8.200000
78   United States, United Kingdom, Canada    6.700000
79  United States, United Kingdom, Germany    6.600000
80    United States, United Kingdom, India    7.800000

[81 rows x 2 columns]

#### Ejercicio 8: ¿Cuántos títulos fueron lanzados en cada año agrupados por tipo ('Movie' o 'TV Show')?

In [53]:
df_conteo = df_Netflix.groupby(['release_year', 'type']).size().reset_index(name='count')
df_conteo


release_year   type  count
0          2014  Movie      1
1          2015  Movie     11
2          2016  Movie     30
3          2017  Movie     56
4          2018  Movie     87
5          2019  Movie    103
6          2020  Movie    169
7          2021  Movie     56

#### Ejercicio 9: ¿Cuál es el país con la mayor cantidad de títulos de clasificación 'TV-MA'?

In [57]:
condicion=df_Netflix['rating'] == 'TV-MA'
df_titulos_por_rating = df_Netflix[condicion]
df_conteo = df_titulos_por_rating.groupby('country').size().reset_index(name='count')
pais_con_mas_titulos = df_conteo.loc[df_conteo['count'].idxmax()]
pais_con_mas_titulos


country    United States
count                121
Name: 46, dtype: object

#### Ejercicio 10. ¿Cuál es el promedio de duración de las películas agrupadas por país?

In [58]:
condicion = df_Netflix['type'] == 'Movie'
df_peliculas=df_Netflix[condicion]
#No se como seguir en este, intuyo que como las duraciones son un string peude estar involucrado regex
#pero no se me ocurre como hacerlo

#### Ejercicio 11: ¿Cuál es el número total de títulos por país y clasificación de audiencia ('rating')?

In [60]:
df_conteo = df_Netflix.groupby(['country', 'rating']).size().reset_index(name='count')
df_conteo


country rating  count
0                                 Argentina  TV-14      3
1                                 Argentina  TV-MA      2
2                  Australia, United States  TV-14      1
3                                   Austria  TV-14      1
4                                   Austria  TV-MA      1
..                                      ...    ...    ...
124           United States, United Kingdom     PG      1
125           United States, United Kingdom  TV-14      1
126   United States, United Kingdom, Canada      R      1
127  United States, United Kingdom, Germany  TV-MA      1
128    United States, United Kingdom, India      R      1

[129 rows x 3 columns]

#### Ejercicio 12: ¿Cuántos títulos fueron lanzados antes del año 2000 agrupados por país?

In [64]:
df_antes_2000 = df_Netflix[df_Netflix['release_year'] < 2000]
df_conteo = df_antes_2000.groupby('country').size().reset_index(name='count')
df_conteo



Empty DataFrame
Columns: [country, count]
Index: []

#### Ejercicio 13: ¿Cuál es la puntuación promedio en IMDb de los títulos lanzados en 2021 agrupados por país?

In [65]:
df_2021 = df_Netflix[df_Netflix['release_year'] == 2021]
df_promemdio = df_2021.groupby('country')['IMDB Score'].mean().reset_index()
df_promedio

type  IMDB Score
0  Movie    6.210916

#### Ejercicio 14: ¿Cuántos títulos fueron dirigidos por cada director? Ordena los resultados de mayor a menor

In [66]:
df_conteo=df_Netflix.groupby('director').size().reset_index(name='count')
df_conteo_ordenado = df_conteo.sort_values(by='count', ascending=False)
df_conteo_ordenado


director  count
285                                                McG      3
78                                   Christopher Guest      2
108                                       David Michôd      2
224                                    Julien Leclercq      2
124                           Elaine McMillion Sheldon      2
..                                                 ...    ...
25                                         Anurag Basu      1
26                                      Anurag Kashyap      1
27   Anurag Kashyap, Dibakar Banerjee, Karan Johar,...      1
28                                         Anvita Dutt      1
3                                           Adrian Teh      1

[462 rows x 2 columns]

#### Ejercicio 15: ¿Cuál es el lenguaje más frecuente en los títulos lanzados en los últimos 5 años agrupado por país?

In [69]:
condicion= df_Netflix['release_year'] >= 2019
df_ultimos5 = df_Netflix[condicion]
df_conteo = df_ultimos5.groupby(['country', 'Language']).size().reset_index(name='count')
indices = df_conteo.groupby('country')['count'].idxmax()
df_mas_frecuente = df_conteo.loc[indices].reset_index(drop=True)
df_mas_frecuente

country          Language  count
0                                Argentina           Spanish      5
1                                  Austria            German      1
2                     Belgium, Netherlands             Dutch      1
3                                   Brazil        Portuguese      6
4                                   Canada           English      3
5           Canada, Hungary, United States           English      1
6                    Canada, United States           English      1
7    Canada, United States, United Kingdom           English      1
8                                    Chile           Spanish      1
9                     China, United States           English      1
10                                  France            French     11
11                         France, Belgium            French      1
12                   France, United States           English      1
13                                 Germany            German      3
14                  Hungary, United States           English      1
15                                   India             Hindi     25
16                    India, United States           English      1
17                               Indonesia        Indonesian      8
18                                   Italy           Italian     11
19                                   Japan          Japanese      3
20                    Japan, United States          Japanese      1
21                                Malaysia             Malay      1
22                                  Mexico           Spanish      5
23                           Mexico, Spain           Spanish      1
24                   Mexico, United States   English/Spanish      1
25                             Netherlands             Dutch      1
26                                 Nigeria           English      1
27                                  Norway         Norwegian      1
28                 Pakistan, United States           English      1
29                             Philippines          Filipino      1
30              Philippines, United States          Filipino      1
31                                  Poland            Polish      2
32                            South Africa           English      3
33                             South Korea            Korean      5
34              South Korea, United States           English      1
35                                   Spain           Spanish      9
36                          Spain, Germany           Spanish      1
37                   Spain, United Kingdom           English      1
38                                  Sweden           Swedish      1
39                   Sweden, United States   English/Swedish      1
40                                Thailand              Thai      2
41                 Thailand, United States      Thia/English      1
42                                  Turkey           Turkish      4
43                          United Kingdom           English      8
44      United Kingdom, Hungary, Australia           English      1
45                   United Kingdom, Italy           English      2
46    United Kingdom, Japan, United States           English      1
47           United Kingdom, United States           English      5
48                           United States           English    142
49                United States, Argentina           English      1
50                   United States, Canada           English      1
51                  United States, Germany           English      1
52                    United States, Japan  English/Japanese      1
53                   United States, Mexico   English/Spanish      2
54           United States, United Kingdom           English      1
55  United States, United Kingdom, Germany           English      1
56    United States, United Kingdom, India           English      1